server/graphql/v2/mutation/PaymentMethodMutations.js

```diff
@@ -50,39 +118,82 @@ const addCreditCard = {
       currency: collective.currency,
       saved: args.isSavedForLater,
       CollectiveId: collective.id,
+      token: args.creditCardInfo.token,
+      data: pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']),
     };
 
-    if (args.creditCardInfo) {
-      const data = pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']);
-      Object.assign(newPaymentMethodData, { token: args.creditCardInfo.token, data });
-    } else if (args.paymentMethod) {
-      Object.assign(newPaymentMethodData, pick(args.paymentMethod, ['data', 'name', 'token']));
-    } else {
-      throw new ValidationFailed('Either creditCardInfo or paymentMethod must be provided');
-    }
-
     let pm = await models.PaymentMethod.create(newPaymentMethodData);
+
     try {
       pm = await setupCreditCard(pm, { collective, user: req.remoteUser });
     } catch (error) {
       if (!error.stripeResponse) {
         throw error;
       }
 
-      // TODO Add support for 3D secure
-      // Currently ignoring the error, which should be ok because we have the `payment.creditcard.confirmation` email.
+      // unsave payment method if saved, we will resave it in confirmCreditCard
+      if (args.isSavedForLater) {
+        await pm.update({ saved: false });
+      }
+
+      pm.stripeError = {
+        message: error.message,
+        account: error.stripeAccount,
+        response: error.stripeResponse,
+      };
+
+      return { paymentMethod: pm, stripeError: pm.stripeError };
     }
 
-    return pm;
+    // Success: delete reference to setupIntent
+    if (pm.data.setupIntent) {
+      delete pm.data.setupIntent;
+    }
+
+    await pm.update({ confirmedAt: new Date(), data: pm.data });
+
+    return { paymentMethod: pm };
+  },
+};
+
+const confirmCreditCard = {
+  type: GraphQLNonNull(CreditCardWithStripeError),
+  description: 'Confirm a credit card is ready for use after strong customer authentication',
+  args: {
+    paymentMethod: {
+      type: new GraphQLNonNull(PaymentMethodReferenceInput),
+    },
+    shouldBeSaved: {
+      type: GraphQLBoolean,
+      description: 'Whether this payment method should be marked as saved for future payments',
+    },
+  },
+  async resolve(_, args, req) {
+    const paymentMethod = await fetchPaymentMethodWithReference(args.paymentMethod);
+
+    if (!paymentMethod || !req.remoteUser.isAdmin(paymentMethod.CollectiveId)) {
```

Comment:
User might be unauthenticated

```suggestion
    if (!paymentMethod || !req.remoteUser?.isAdmin(paymentMethod.CollectiveId)) {
```